foreshadow/core/fs_logging.py

```diff
@@ -0,0 +1,264 @@
+
+    Args:
+        level: The corresponding log function level. E.g. 'debug'
+
+    Returns:
+        The log function callable.
+
+    Raises:
+        ValueError: if logging level does not have a logging function.
+
+    """
+    log_fn = getattr(_this_module, level, None)
+    if log_fn is None:
+        raise ValueError("level: '{}' is not a valid logging level.")
+    return log_fn
+
+
+class SyncWrite(object):
+    """Single-instance object to handle buffered writing to file."""
+
+    # TODO turn this into synchronous Queue based writing.
+    def __init__(
+        self, buffer_size=100, outfile="gui_data.txt", overwrite=True
+    ):
+        super(SyncWrite, self).__init__()
+        self.threads = []
+        self._buffer_size = buffer_size
+        self.outfile = outfile
+        self.first_write = overwrite
+        self.buffer = []
+        self.writing = False
+
+    @property
+    def buffer_size(self):
+        """How long to wait before writing.
+
+        Returns:
+            buffer size as integer
+
+        """
+        return self._buffer_size
+
+    @buffer_size.setter
+    def buffer_size(self, value):
+        """Set buffer size to value and ensure it makes sense.
+
+        Args:
+            value: integer to set buffer_size to.
+
+        """
+        if value < 1:
+            log_fn = _get_log_fn("warning")
+            log_fn("trying to set buffer size < 1. Automatically setting to 1")
+            value = 1
+        if value < len(self.buffer):
+            log_fn = _get_log_fn("warning")
+            log_fn(
+                "trying to set buffer < current buffer length. "
+                "Automatically writing buffer to file before setting."
+            )
+            self.write()
+        self._buffer_size = value
+
+    def write(self):
+        """Write buffer (gui calls) to file."""
+        log_fn = _get_log_fn("debug")
+        log_fn("writing gui buffer with: %d msgs to file" % len(self.buffer))
+        io_type = "w" if self.first_write else "a"
+        with open(self.outfile, io_type) as outfile:
+            for msg in self.buffer:
+                outfile.write(msg)
+                outfile.write("\n")
+        self.first_write = False
+        self.buffer = []
+
+    def __call__(self, schema, details):
+        """Write to Gui. Will store in buffer until it reaches buffer_size.
+
+        Args:
+              schema: schema from foreshadow.core.gui to serialize
+              details: dict of values corresponding to schema.
+
+        """
+        msg = str(schema.load(details).data)
+        # perform write operation to some nosql database or file
+        if len(self.buffer) >= self.buffer_size - 1:  # -1 as append.
+            self.buffer.append(msg)
+            self.write()
+        else:
+            self.buffer.append(msg)
+
+
+gui_fn = SyncWrite()
+
+
+def set_level(level_name):
+    """Set logger level.
+
+    Args:
+        level_name: level to set logger to. Must be in LEVELS.
```

Comment:
might be worth linking LEVELS using sphinx attribute linking and writing documentation on the different levels supported on the module attribute.

Reply:
Will ask you about this tomorrow.